packages/core/lib/commands/call/run.js

```diff
@@ -0,0 +1,167 @@
+module.exports = async function (options) {
+  const fs = require("fs");
+  const util = require("util");
+  const { Environment } = require("@truffle/environment");
+  const OS = require("os");
+  const Codec = require("@truffle/codec");
+  const Encoder = require("@truffle/encoder");
+  const Decoder = require("@truffle/decoder");
+  const TruffleError = require("@truffle/error");
+  const { fetchAndCompile } = require("@truffle/fetch-and-compile");
+  const loadConfig = require("../../loadConfig");
+
+  if (options.url && options.network) {
+    const message =
+      "" +
+      "Mutually exclusive options, --url and --network detected!" +
+      OS.EOL +
+      "Please use either --url or --network and try again." +
+      OS.EOL +
+      "See: https://trufflesuite.com/docs/truffle/reference/truffle-commands/#call" +
+      OS.EOL;
+    throw new TruffleError(message);
+  }
+
+  let config = loadConfig(options);
+  await Environment.detect(config);
+
+  const [contractNameOrAddress, functionNameOrSignature, ...args] = config._;
+  let functionEntry, transaction;
+
+  const { encoder, decoder } = config.fetchExternal
+    ? await sourceFromExternal(contractNameOrAddress, config)
+    : await sourceFromLocal(contractNameOrAddress, config);
+
+  ({ abi: functionEntry, tx: transaction } = await encoder.encodeTransaction(
+    functionNameOrSignature,
+    args
+  ));
+
+  if (functionEntry.stateMutability !== "view") {
+    console.log(
+      `WARNING!!! Not a view function\n  ABI stateMutability: ${functionEntry.stateMutability}`
+    );
+  }
+
+  const provider = new Encoder.ProviderAdapter(config.provider);
+  let result;
+  try {
+    result = await provider.request({
```

Comment:
Btw, another reason to do this is that `request` returns `Promise<any>`.  That's dangerous!  You should use something better-typed than that.

Reply:
`call` can be added as an adapter method but why is that more preferable than a raw request?

Reply:
Because that's the point of the adapter!  The whole point of the adapter is to contain reusable methods that other components can -- in contexts where for whatever reason we don't want to use web3 or ethers -- use to perform RPC methods without having to muck about with raw requests.  Using the adapter just so you can perform a request yourself, rather than calling a convenient method to have it construct and perform the request for you, defeats the point of it!

Alternatively, rather than use the adapter at all, maybe it's acceptable to just use web3 here.  I assumed that you used the adapter because after talking to @gnidan you concluded there was some reason not to use web3 here.  But I'm not actually seeing such a reason offhand?  So likely you could just use web3 and spare yourself some trouble.

Reply:
I gave it a thought and I think it might be a better idea to add `call` as an adapter method so that it can be reused. I am going to try that. Thanks @haltman-at .